src/TorchSharp/Tensor/Storage.cs

```diff
@@ -0,0 +1,522 @@
+                    ValidateNotScalar();
+                    CheckIndex(index);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        return ptr[index];
+                    }
+                }
+                set {
+                    ValidateNotScalar();
+                    CheckIndex(index);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        ptr[index] = value;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Accesses a slice of a storage instance.
+            /// </summary>
+            /// <param name="range">The range, expressed as a tuple [start,end)</param>
+            public Storage<T> this[(int? start, int? end) range] {
+                get {
+                    ValidateNotScalar();
+                    var start = CheckIndex(range.start.HasValue? range.start.Value : 0);
+                    var end = CheckIndex(range.end.HasValue ? range.end.Value : Count);
+                    var count = end - start;
+                    unsafe {
+                        var ptr = _tensor_data_ptr + sizeof(T) * start;
+                        return new Storage<T>(ptr, _tensor.dtype, count);
+                    }
+                }
+                set {
+                    ValidateNotScalar();
+                    if (value._tensor is not null || value._tensor_data_ptr != IntPtr.Zero)
+                        throw new InvalidOperationException();
+                    var v = value._scalarValue;
+                    var start = CheckIndex(range.start.HasValue ? range.start.Value : 0);
+                    var end = CheckIndex(range.end.HasValue ? range.end.Value : Count);
+                    unsafe {
+                        T* ptr = (T*)_tensor_data_ptr;
+                        for (int i = start; i < end; i++)
+                            ptr[i] = v;
+                    }
+                }
+            }
+
+#if !NETSTANDARD2_0_OR_GREATER
+            /// <summary>
+            /// Accesses a slice of a storage instance.
+            /// </summary>
+            /// <param name="range">The range.</param>
+            public Storage<T> this[System.Range range] {
+                get {
+                    ValidateNotScalar();
+                    var start = CheckIndex(range.Start);
+                    var end = CheckIndex(range.End);
+                    return this[(start, end)];
+                }
+                set {
+                    ValidateNotScalar();
+                    var start = CheckIndex(range.Start);
+                    var end = CheckIndex(range.End);
+                    this[(start, end)] = value;
+                }
+            }
+
+            private int CheckIndex(System.Index index)
+            {
+                var i = index.IsFromEnd ? Count - index.Value : index.Value;
+                if (i < 0 || i >= Count)
```

Comment:
>if (i < 0 || i >= Count) [](http://example.com/codeflow?start=16&length=24)

nit: sometimes we optimize conditions like this by doing it as 

```C#
if ((uint)i >= Count)
    ...
```

Reply:
Unless that's a big performance deal, I'd rather not. It's clever, but less readable.

Reply:
This would cause an `OverflowException` when the calling code is in checked mode. ie. 

```csharp
bool TestCondition(int i, int count) => (uint)i >= count;

checked
{
	return TestCondition(-5, 3); // throws: OverfowException; Arithmetic operation resulted in an overflow.
}
```

Reply:
@MovGP0 FYI you are wrong. `checked` affects only arithmetic operators inside the `checked`, it does not affect operators inside functions you are calling from `checked` block.